### ISSUE TITLE
[API Proposal]: IRandom interface (which Random and RandomNumberGenerator implement)

### ISSUE THREAD

**Issue by BenMakesGames** [dotnet/runtime#73567]
### Background and motivation

(I _did_ find this: https://github.com/dotnet/runtime/issues/28229 and I think I can provide some more details and use-cases than were provided there.)

Bring `RandomNumberGenerator` and `Random` under a single interface (`IRandom`), in order to:
* allow capabilities (such as "give me a random int") to be shared between these implementations (extensions on `IRandom`)
* allow third-party packages to easily add new capabilities to all random number generators
* enable custom random and pseudo-random number generators (such as perlin noise generators), without losing the methods provided to `IRandom`
* enable easier testing, by facilitating mocking, and the use of `IRandom` as a dependency

### API Proposal

```csharp
namespace System;

// The heart of any random-number generator is in generating some bytes;
// all other methods can be derived from this.
public interface IRandom
{
    void GetBytes (byte[] data);
    void GetBytes (Span<byte> data);
    byte[] GetBytes (int count);
    void GetBytes (byte[] data, int offset, int count);
    // ^ Can some of these be moved into IRandomExtensions? (below)
}
```

```csharp
namespace System;

public class Random: IRandom
{
    ...
}
```

```csharp
namespace System.Security.Cryptography;

public class RandomNumberGenerator: IRandom, IDisposable
{
    // Not sure if it's feasible to make RandomNumberGenerator's constructor public;
    // my gut is to say "it might be nice if it were", but I can imagine that there are good reasons it isn't.
    ...
}
```

```csharp
namespace System; // Somewhere else?

public static class IRandomExtensions
{
    public static int Next(this IRandom random) { ... }
    public static int Next(this IRandom random, int maxValue) { ... }
    public static int Next(this IRandom random, int minValue, int maxValue) { ... }
    public static long NextInt64(this IRandom random) { ... }
    public static long NextInt64(this IRandom random, long maxValue) { ... }
    public static long NextInt64(this IRandom random, long minValue, long maxValue) { ... }
    public static float NextSingle(this IRandom random) { ... }
    public static double NextDouble(this IRandom random) { ... }
}
```

### API Usage

```csharp
public class MyService: IMyService
{
    private IRandom Rng { get; }

    public MyService(IRandom rng)
    {
        Rng = rng;
    }

    public MyProcGenContent GenerateContent(int x, int y)
    {
        ...
    }
}
```

```csharp
public class MyPerlinNoiseGenerator: IRandom
{
    ...
}
```

```csharp
public class MyIntentionallyBad8BitGamingConsoleInspiredRng: IRandom
{
    ...
}
```

### Alternative Designs

I haven't used default interface methods much, but I assume the methods defined in `IRandomExtensions` could be implemented as those, instead? Not sure about the pros and cons of doing it that way vs. as "traditional" interface extensions.

### Risks

I don't have an intimate knowledge of `Random`, or, especially, of `RandomNumberGenerator`. I _think_ the proposed changes can be made without introducing breaking changes, but glancing through the source, I see thread-aware code, and other things in there that I'm not familiar with.

**Comment by MichalPetryka** [comment]
Methods in `Random` are explicitly virtual so that you can derive from it and provide your own implementation, you can thus already make one that wraps RandomNumberGenerator with no BCL changes.

**Comment by bartonjs** [comment]
In general, we don't like interfaces that look like classes, we only like interfaces that describe capability (e.g. IEnumer**able**), and IRandom fails on that front.

The CSPRNG (System.Security.Cryptography.RandomNumberGenerator) has moved to static methods to indicate that deriving the type or otherwise substituting it isn't generally a good idea; and no one who needs to use the CSPRNG would benefit from an interface here.  So RandomNumberGenerator should be left out.

That now leaves the proposal as System.Random and System.IRandom; but as everything on System.Random is virtual, the interface doesn't really have a lot of inherent value.  Combined with the first point, I don't think it's a good idea.

**Comment by BenMakesGames** [comment]
I would describe the capability as "the ability to generate a byte (or, for convenience, many bytes)", and that's all I think `IRandom` should expose. the means by which these bytes are generated will vary based on implementation: indexable noise functions, forward-only & repeatable algorithms like what `Random` uses, and however the cryptographic `RandomNumberGenerator` works (I assume it's forward-only, and non-repeatable, like reading pixels from a video feed of some lava lamps :P)

I propose placing all the other methods (ex: `NextDouble`) as extensions for `IRandom`; not on `IRandom` itself.

HOWEVER: in explaining this, I am realizing that my proposal does not enable indexable RNG, like perlin noise. I'm happy to tinker more (find & polish up some actual C#) and update the proposal, if there's interest.

also, if involving the cryptographic `RandomNumberGenerator` raises concerns & increases complexity, I'm not adverse to excluding that; I saw it as an "easy add", but it sounds like I may have been wrong to think so.

**Comment by bartonjs** [comment]
Well, "the ability to generate a byte (or, for convenience, many bytes)" would be `IByteGenerator`, not `IRandom`.

It's a little unfortunate that there's not a good "funneling" method on System.Random, but your PerlinNoise RNG can easily be

```C#
public class PerlinNoiseGenerator : System.Random
{
    public override int Next()
    {
        int ret = 0;
        NextBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref ret, 1)));
        return ret;
    }

    public override int Next(int maxValue)
    {
        // ...
    }

    public override int Next(int minValue, int maxValue)
    {
        // ...
    }

    public override long NextInt64()
    {
        long ret = 0;
        NextBytes(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref ret, 1)));
        return ret;
    }

    public override long NextInt64(long maxValue)
    {
        // ...
    }

    public override long NextInt64(long minValue, long maxValue)
    {
        // ...
    }

    public override void NextBytes(byte[] buffer)
    {
        ArgumentNullException.ThrowIfNull(buffer);
        NextBytes(buffer.AsSpan());
    }

    public override void NextBytes(Span<byte> buffer)
    {
        // Extract the next sequence from the Perlin algorithm
        // ...
    }

    public override float NextSingle()
    {
        return Next() * 1.0f / int.MaxValue;
    }

    public override double NextDouble()
    {
        return NextInt64() * 1.0 / long.MaxValue;
    }

    protected override double Sample()
    {
        return NextDouble();
    }
}
```

And if the only reason the interface is showing up is for DI reasons, you can use classes as the key in the DI registration methods. (e.g. instead of `whatever.AddSingleton<IRandom, PerlinNoiseGenerator>();` you can call `whatever.AddSingleton<Random, PerlinNoiseGenerator>();` and that'll create an instance of the Perlin generator instead of the normal System.Random ctor whereever a System.Random is needed)

**Comment by BenMakesGames** [comment]
you had me all excited that `IByteGenerator` was a thing, and I went googling :P

I liked your example on how to treat Random as an 'interface'. that's cool to know, however it doesn't seem to be very discoverable (my ignorance does not seem to be uncommon): I rarely see a third-party package which ask the developer to pass in an instance of `Random` (instead they `new` up an instance themselves, internally, ex: https://github.com/StefH/RandomDataGenerator/blob/master/src/RandomDataGenerator/Generators/RandomStringGenerator.cs ); similarly, I don't think I've _ever_ seen a third-party class extend `Random` in order to act as a drop-in replacement (instead inventing some new class or collection of methods which is not compatible with anything else).

searching nuget for "random" just now, I was amused to see the first result is for some kind of password generation library with the text "Removed usage of RNG Crypto Provider and used Random Number Generator" directly in the search results. if the code had asked for a random number generator, instead of depending on a concrete implementation, a dev could have upgraded their use of the library before the package maintainer did.

I think even casual devs know the rule of thumb that using an interface is often better than a concrete classes (IDEs push this, too, ex: by recommending `ICollection` or `IEnumerable` instead of `List`), so having an explicit interface might help encourage writing more reusable code.

tl;dr: `Random` _can_ be used as a service, as it is currently implemented, but this fact does not seem to be very discoverable; few people appear to be writing code that takes advantage of this fact. use of an explicit interface may help foster better code.

**Comment by bartonjs** [comment]
> I think even casual devs know the rule of thumb that using an interface is often better than a concrete classes

The .NET design guidelines say the opposite, in fact. To quote Framework Design Guidelines, 3rd edition, p97 (section 4.3, "Choosing Between Class and Interface")

> DO favor defining classes over interfaces.

which continues on the next page to

> DO use abstract classes instead of interfaces to decouple the contract from implementations.

While ASP.NET and other higher level layers may like interfaces, at the bottom level we discourage using them unless it's an absolute slam-dunk that they're a) needed and b) useful.  (Many of the API Reviews that we live stream at https://www.youtube.com/c/NETFoundation/search?query=github%20quick%20reviews will show many cases where a proposal came in with an interface... and it didn't leave with one)

Basically, the more "useful" an interface looks, or the more it looks like a class with no implementation, the less appropriate it is as an interface.  `IDisposable` is an utterly useless interface on its own, but it conveys that an object has (or is concerned it may one day have) state that needs to be cleaned up.  You'd never new up an `IEnumerable`, it just says "I'm capable of providing some values, one after the other".  Conversely, `ICryptoTransform` kinda looks like a class... and when we added Span to .NET we found we couldn't make ICryptoTransform play along.

> "Removed usage of RNG Crypto Provider and used Random Number Generator" 

That's an example of where our guidance for the CSPRNG has changed.  From that description, they changed something like

```C#
using (RandomNumberGenerator rng = new RNGCryptoServiceProvider())
{
   rng.GetBytes(buffer);
}
```

to

```C#
RandomNumberGenerator.Fill(buffer);
```

The update there is a) stop allocating wasteful instances and b) we think this concept should never have been polymorphic.  If we had an IRandom, it would have been "stop using IRandom, use RandomNumberGenerator.Fill directly"

> IDEs push this, too, ex: by recommending ICollection or IEnumerable instead of List

Once an interface was justified (or we just didn't learn until after adding it that adding it was a bad idea... c.f. IList vs IReadOnlyList) then it's correct to make parameters accept the interface when practical, because it's more flexible.

**Comment by BenMakesGames** [comment]
alrighty. thanks for talking this stuff through with me. I've learned some things, and become confused about others, but that's fine, too :P I'll close this proposal; maybe I'll be back someday with a new one :P
